src/perl/bin/sample_udf_updater

```diff
@@ -0,0 +1,240 @@
+#! /usr/bin/env perl
+
+package sample_udf_updater;
+
+use strict;
+use warnings;
+use FindBin qw($Bin);
+use lib ( -d "$Bin/../lib/perl5" ? "$Bin/../lib/perl5" : "$Bin/../lib" );
+
+use Moose;
+use Carp;
+use Readonly;
+use XML::LibXML;
+use List::MoreUtils qw/uniq/;
+
+use wtsi_clarity::clarity::process;
+
+with qw/wtsi_clarity::util::configurable
+        wtsi_clarity::util::roles::clarity_request
+        wtsi_clarity::util::clarity_elements/;
+
+our $VERSION = '0.0';
+
+## no critic(ValuesAndExpressions::RequireInterpolationOfMetachars)
+Readonly::Scalar my $SAMPLE_URI_PATH  => q{/art:details/art:artifact/sample/@uri};
+## use critic
+Readonly::Scalar my $SAMPLE_PATH      => q{/smp:details/smp:sample};
+
+has 'process_url'  => (
+  isa             => 'Str',
+  is              => 'ro',
+  required        => 1,
+);
+
+has 'process_doc'  => (
+  isa             => 'wtsi_clarity::clarity::process',
+  is              => 'ro',
+  required        => 0,
+  traits          => [ 'NoGetopt' ],
+  lazy_build      => 1,
+  handles         => {'findnodes' => 'findnodes'},
+);
+sub _build_process_doc {
+  my ($self) = @_;
+  my $xml = $self->fetch_and_parse($self->process_url);
+  return wtsi_clarity::clarity::process->new(xml => $xml, parent => $self);
+}
+
+has 'udf_name'  => (
+  isa             => 'Str',
+  is              => 'ro',
+  required        => 1,
+);
+
+has 'udf_value'  => (
+  isa             => 'Str',
+  is              => 'ro',
+  required        => 1,
+);
+
+has '_input_artifacts' => (
+  isa             => 'XML::LibXML::Document',
+  is              => 'rw',
+  required        => 0,
+  lazy_build      => 1,
+);
+sub _build__input_artifacts {
+  my $self = shift;
+
+  return $self->process_doc->input_artifacts;
+}
+
+has '_samples' => (
+  isa             => 'XML::LibXML::Document',
+  is              => 'rw',
+  required        => 0,
+  lazy_build      => 1,
+);
+sub _build__samples {
+  my $self = shift;
+
+  my @sample_uri_node_list = $self->_input_artifacts->findnodes($SAMPLE_URI_PATH)->get_nodelist();
+  my @sample_uris = map { $_->getValue } @sample_uri_node_list;
+
+  return $self->request->batch_retrieve('samples', \@sample_uris);
+}
+
+sub _update_samples_with_udfs {
+  my $self = shift;
+
+  my @sample_node_list = $self->_samples->findnodes($SAMPLE_PATH)->get_nodelist();
+
+  foreach my $sample (@sample_node_list) {
+    $self->_add_udf_to_sample($sample);
+  }
+
+  return;
+}
+
+sub _add_udf_to_sample {
```

Comment:
I think there's already a couple of methods you can use from the `clarity_elements` role that'll do this for you